# Exercise 2.A Fundamentals of pandas

In [20]:
#importing libraries
import numpy as np
import pandas as pd

## Series

In [23]:
#creating 4 new variables
aa = [1,2,3,4] #list
ab = (5,6,7,8) #tuple
ac = {9,10,11,12} #set
ad = {'first':13,'second':14,'third':15,'fourth':16} #dictionary

#printing outputs
print("aa:",aa)
print("ab:",ab)
print("ac:",ac)
print("ad:",ad)

aa: [1, 2, 3, 4]
ab: (5, 6, 7, 8)
ac: {9, 10, 11, 12}
ad: {'first': 13, 'second': 14, 'third': 15, 'fourth': 16}


In [25]:
#creating 4 more new variables

#assigning series to previous variables
ser1 = pd.Series(aa)
ser2 = pd.Series(ab)
ser3 = pd.Series(ac)
ser4 = pd.Series(ad)

#printing output
print("ser1: \n",ser1)
print("\nser2: \n",ser2)
print("\nser3: \n",ser3) #TypeError: 'set' type is unordered
print("\nser4: \n",ser4)

TypeError: 'set' type is unordered

In [27]:
#creating a new variable
ser5 = pd.Series(17,18,19,20) #literal values

#TypeError: Index(...) must be called with a collection of some kind, 18 was passed

TypeError: Index(...) must be called with a collection of some kind, 18 was passed

## DataFrames from Series

In [30]:
#creating new variable for DataFrame1 using ser1
df1 = pd.DataFrame(ser1)
df1 #displaying df1

,0
0,1
1,2
2,3
3,4


In [32]:
#adding two columns to df1

#creating ser3 to be usable
ser3 = pd.Series(list(ac))

df1.insert(1,"col2",ser2) #inserting ser2 as column
df1.insert(2,"col3",ser3) #inserting ser3 as column

#displaying new df1
df1
#NameError: name 'ser3' is not defined. creating a ser3 that is usable ^

,0,col2,col3
0,1,5,9
1,2,6,10
2,3,7,11
3,4,8,12


In [34]:
# adding third column (ser4)

#creating ser4 to be usable
ser4 = pd.Series(ad)

df1.insert(3,"col4",ser4)
df1 #displaying output
#NameError: name'ser4' is not defined. creaint a ser4 that is usable ^

#ser4 has dictionary keys as its index, there are no numeric values in place.
#df1 uses numeric indexes (0-3), so pandas cannot align them.
#This caused the new column to be filled with NaN.

,0,col2,col3,col4
0,1,5,9,NaN
1,2,6,10,NaN
2,3,7,11,NaN
3,4,8,12,NaN


In [36]:
#adding a new forth row
df1.loc[4] = [5,9,12,0] #using loc to add new row
df1 #displaying output

,0,col2,col3,col4
0,1,5,9,NaN
1,2,6,10,NaN
2,3,7,11,NaN
3,4,8,12,NaN
4,5,9,12,0.0


## Reading Data into a DataFrame

In [41]:
#creating new variable df2
df2 = pd.read_csv("sakila_example.csv") #reading sakila csv
df2 #displaying output

,title,description,category,language,actor_name
0,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,PENELOPE GUINESS
1,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,CHRISTIAN GABLE
2,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,LUCILLE TRACY
3,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,SANDRA PECK
4,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,JOHNNY CAGE
...,...,...,...,...,...
257,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,GARY PENN
258,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,CHRISTOPHER WEST
259,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,DUSTIN TAUTOU
260,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,ANGELA HUDSON


In [43]:
#displaying the top 5 records
df2.head()

,title,description,category,language,actor_name
0,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,PENELOPE GUINESS
1,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,CHRISTIAN GABLE
2,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,LUCILLE TRACY
3,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,SANDRA PECK
4,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,JOHNNY CAGE


In [45]:
#displaying the last 10 records
df2.tail(10)

,title,description,category,language,actor_name
252,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,MILLA KEITEL
253,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,GROUCHO DUNST
254,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,BURT TEMPLE
255,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,UMA WOOD
256,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,RIP WINSLET
257,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,GARY PENN
258,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,CHRISTOPHER WEST
259,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,DUSTIN TAUTOU
260,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,ANGELA HUDSON
261,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,JAMES PITT


In [47]:
#displaying information summary
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 262 entries, 0 to 261
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   title        262 non-null    object
 1   description  262 non-null    object
 2   category     262 non-null    object
 3   language     262 non-null    object
 4   actor_name   262 non-null    object
dtypes: object(5)
memory usage: 10.4+ KB


### Summary Info for df2
This dataset had 262 rows & 5 columns. All data types are objects, meaning it is text-based.
The columns are named title, description, category, language and actor_name.
Every column has 262 non-null values, so there are no missing values in this dataset.
There is 10.4 KB memory stored, so this tells how large or little the dataset is.

In [50]:
#showing column statistics
df2.describe() #using describe to show column statistics

,title,description,category,language,actor_name
count,262,262,262,262,262
unique,46,46,15,1,145
top,ARABIA DOGMA,A Touching Epistle of a Madman And a Mad Cow w...,Horror,English,OPRAH KILMER
freq,12,12,55,262,8


### Column Statistics for df2
Because all the columns are text, describe was able to show the count, unique, top, frequent values.
This is useful when trying to see which columns appear most in the dataset.
It is not useful when attempting a deeper analysis since the columns are text and not numeric.

In [53]:
#creating new variable
df3 = pd.read_csv("northwind_example.csv") #reading csv file
df3 #displaying output

,OrderID,ProductName,UnitPrice,Quantity
0,10248,Queso Cabrales,14.0,12
1,10248,Singaporean Hokkien Fried Mee,9.8,10
2,10248,Mozzarella di Giovanni,34.8,5
3,10249,Manjimup Dried Apples,42.4,40
4,10249,Tofu,18.6,9
...,...,...,...,...
495,10435,Mozzarella di Giovanni,27.8,10
496,10436,Gnocchi di nonna Alice,30.4,40
497,10436,Wimmers gute Semmelkndel,26.6,30
498,10436,Rhnbru Klosterbier,6.2,24


In [55]:
#showing column statistics
df3.describe() #using describe to show column statistics

,OrderID,UnitPrice,Quantity
count,500.000000,500.000000,500.000000
mean,10341.364000,23.371400,24.330000
std,54.442663,27.334436,18.069702
min,10248.000000,2.000000,1.000000
25%,10294.000000,10.400000,10.000000
50%,10341.000000,15.800000,20.000000
75%,10389.000000,27.800000,32.000000
max,10436.000000,210.800000,120.000000


### Column Statistics for df3
Using describe was helpful because this dataset has numeric columns.
Because of this, we are able to see certain stats like min, max and mean, letting us see how the numbers are spread out.
This is different from df2, because df2 had text columns and using describe only showed unique values. With df3
we are able to gain more insight since the dataset is numeric, which lets us see more statistical information.

## Coding a DataFrame

In [59]:
#creating a new variable as dictionary
test_data = {'ID':[1,2,3,4,5,6,7,8],
             'Name':['Alicia','Bob','Charlie','Dominic','Eve','Frank','Grace','Heidi'],
             'Age':[25,np.nan,22,19,31,35,np.nan,18],
             'City':['NY','LA','Chicago','Houston','Phoenix','Boston','Austin','San Diego'],
             'Store':[85,92,-1,65,78,np.nan,55,90]}

In [61]:
#creating a DataFrame
df4 = pd.DataFrame(test_data) #using test_data to create DF
df4 #displaying output

,ID,Name,Age,City,Store
0,1,Alicia,25.0,NY,85.0
1,2,Bob,NaN,LA,92.0
2,3,Charlie,22.0,Chicago,-1.0
3,4,Dominic,19.0,Houston,65.0
4,5,Eve,31.0,Phoenix,78.0
5,6,Frank,35.0,Boston,NaN
6,7,Grace,NaN,Austin,55.0
7,8,Heidi,18.0,San Diego,90.0


In [63]:
#exporting df4 to CSV file
df4.to_csv("test_data.csv",index=False)